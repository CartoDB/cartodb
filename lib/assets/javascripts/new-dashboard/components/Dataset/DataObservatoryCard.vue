<template>
  <div class="dataobservatory-card" @click="toggleExpand">
    <div class="u-flex">
      <div class="dataobservatory-type u-mr--16" :class="`dataobservatory-type--${categorySlug}`">
      </div>
      <div class="dataobservatory-information">
        <h3 class="title is-body is-txtGrey">{{$props.dataset.title}}</h3>
      </div>
      <div class="dataobservatory-morebutton">
        <a class="text is-semibold is-small">{{buttonText}}</a>
      </div>
    </div>

    <div class="u-ml--64 dataobservatory-extra-information" :style="`max-height: ${elemHeight}`" :class="{'is-open': isOpen}" ref="expandableElement">
      <p class="text is-caption u-mt--16 u-mb--36">{{$props.dataset.description}}</p>
      <div class="grid grid--align-center">
        <a href="#" class="title is-small u-mr--24 button" @click.stop>{{$t('dataObservatory.viewSample')}}</a>
        <a href="#" class="title is-small button button--outline" @click.stop>{{$t('dataObservatory.interested')}}</a>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'DataObservatoryCard',
  data: function () {
    return {
      isOpen: false
    };
  },
  props: {
    dataset: Object
  },
  computed: {
    categorySlug () {
      let category = this.$props.dataset.category;
      return category.toLowerCase().replace(/\s+/g, '-');
    },
    buttonText () {
      return this.isOpen ? this.$t('dataObservatory.close') : this.$t('dataObservatory.viewMore');
    },
    maxExpandableHeight () {
      const realHeight = this.$refs.expandableElement.scrollHeight;
      return `${realHeight}px`;
    },
    elemHeight () {
      if (this.isOpen) {
        return this.maxExpandableHeight;
      } else {
        return '0';
      }
    }
  },
  methods: {
    toggleExpand () {
      this.isOpen = !this.isOpen;
    }
  }
};
</script>

<style scoped lang="scss">
@import 'new-dashboard/styles/variables';

.dataobservatory-card {
  padding: 22px 36px 22px 24px;
  border: 1px solid $neutral--300;
  border-radius: 4px;
  cursor: pointer;
}

.dataobservatory-information {
  display: flex;
  flex-grow: 1;
  align-items: center;
}

.dataobservatory-morebutton {
  flex-shrink: 0;
  align-self: center;
  width: 64px;
  text-align: right;
}

.dataobservatory-type {
  flex-shrink: 0;
  width: 48px;
  height: 48px;
  border-radius: 2px;
  background-repeat: no-repeat;

  &--financial {
    background-image: url("../../assets/icons/datasets/data-observatory/data-types/finance.svg");
  }
  &--demographics {
    background-image: url("../../assets/icons/datasets/data-observatory/data-types/finance.svg"); //CHANGE
  }
}

.dataobservatory-extra-information {
  // max-height: 0;
  overflow: hidden;
  transition: all 0.3s cubic-bezier(0.4, 0.01, 0.165, 0.99);

  // &.is-open {
  //   max-height: 100%;
  // }
}

</style>
