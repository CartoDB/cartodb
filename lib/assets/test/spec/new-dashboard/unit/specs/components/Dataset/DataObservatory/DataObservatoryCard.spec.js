import { shallowMount } from '@vue/test-utils';
import DataObservatoryCard from 'new-dashboard/components/Dataset/DataObservatoryCard';

const dataObservatorySample = [
  {
    'title': 'Data 1',
    'category': 'Category 1',
    'description': 'This is a description test',
    'geographies': ['Country 1'],
    'companies': ['Company 1']
  },
  {
    'title': 'Data 2',
    'category': 'Category 2',
    'description': 'This is another description test',
    'geographies': ['Country 1', 'Country 2'],
    'companies': ['Company 2', 'Company 3']
  },
  {
    'title': 'Test 3',
    'category': 'Category 2',
    'description': 'This is another description test',
    'geographies': ['Country 4', 'Country 2'],
    'companies': ['Company 1', 'Company 3']
  }
];

const $t = key => key;

describe('DataObservatoryCard.vue', () => {
  it('should render correct contents', () => {
    const dataObservatoryCard = shallowMount(DataObservatoryCard, {
      mocks: {
        $t
      },
      propsData: {
        dataset: dataObservatorySample[0]
      }
    });

    expect(dataObservatoryCard).toMatchSnapshot();
  });

  it('Should open the card', () => {
    const dataObservatoryCard = shallowMount(DataObservatoryCard, {
      mocks: {
        $t
      },
      propsData: {
        dataset: dataObservatorySample[0]
      }
    });

    expect(dataObservatoryCard.vm.isOpen).toBe(false);
    const card = dataObservatoryCard.find('.dataobservatory-card');
    card.trigger('click');

    expect(dataObservatoryCard).toMatchSnapshot();
    expect(dataObservatoryCard.vm.isOpen).toBe(true);

    card.trigger('click');
    expect(dataObservatoryCard.vm.isOpen).toBe(false);
  });

  describe('Computed', () => {
    it('Should return the correct text for the button', () => {
      const dataObservatoryCard = shallowMount(DataObservatoryCard, {
        mocks: {
          $t
        },
        propsData: {
          dataset: dataObservatorySample[0]
        }
      });

      expect(dataObservatoryCard.vm.buttonText).toBe('dataObservatory.viewMore');

      const card = dataObservatoryCard.find('.dataobservatory-card');
      card.trigger('click');

      expect(dataObservatoryCard.vm.buttonText).toBe('dataObservatory.close');
    });
  });
});
