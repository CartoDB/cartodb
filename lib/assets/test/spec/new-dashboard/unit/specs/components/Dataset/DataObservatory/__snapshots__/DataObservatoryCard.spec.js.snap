// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataObservatoryCard.vue Should open the card 1`] = `
<div class="dataobservatory-card">
  <div class="u-flex">
    <div class="dataobservatory-type u-mr--16 dataobservatory-type--category-1"></div>
    <div class="dataobservatory-information">
      <h3 class="title is-body is-txtGrey">Data 1</h3>
    </div>
    <div class="dataobservatory-morebutton"><a class="text is-semibold is-small">dataObservatory.close</a></div>
  </div>
  <div class="u-ml--64 dataobservatory-extra-information is-open" style="max-height: 0px;">
    <p class="text is-caption u-mt--16 u-mb--36">This is a description test</p>
    <div class="grid grid--align-center"><a href="#" class="title is-small u-mr--24 button">dataObservatory.viewSample</a> <a href="#" class="title is-small button button--outline">dataObservatory.interested</a></div>
  </div>
</div>
`;

exports[`DataObservatoryCard.vue should render correct contents 1`] = `
<div class="dataobservatory-card">
  <div class="u-flex">
    <div class="dataobservatory-type u-mr--16 dataobservatory-type--category-1"></div>
    <div class="dataobservatory-information">
      <h3 class="title is-body is-txtGrey">Data 1</h3>
    </div>
    <div class="dataobservatory-morebutton"><a class="text is-semibold is-small">dataObservatory.viewMore</a></div>
  </div>
  <div class="u-ml--64 dataobservatory-extra-information" style="max-height: 0;">
    <p class="text is-caption u-mt--16 u-mb--36">This is a description test</p>
    <div class="grid grid--align-center"><a href="#" class="title is-small u-mr--24 button">dataObservatory.viewSample</a> <a href="#" class="title is-small button button--outline">dataObservatory.interested</a></div>
  </div>
</div>
`;
